### PR TITLE
adding an option for supporting multiple databases

### PR DESCRIPTION
**Pull request by kajatiger**
In [this issue](https://github.com/makandra/geordi/issues/101) I am explaining the problem of dumple not finding the right key for the database in a database.yml that has multiple databases. So here is a solution where we like to offer another option to define that key.

### CODE REVIEW

**Comment by kajatiger** [comment]
I think the support for multiple local databases during dump loading would be nice indeed, but in another PR, since it would add another layer of logic.

File: features/dump.feature [hunk 1]
```diff
@@ -32,3 +36,15 @@ Feature: The dump command
...
+    Then the output should contain "# Dumping the database of staging"
+      And the output should contain "> Connecting to www.example.com"
+      And the output should contain "Util.system! ssh, user@www.example.com, -t, cd /var/www/example.com/current && bash --login -c 'dumple staging --for_download'"
+      And the output should contain "> Downloading remote dump to tmp/staging.dump"
+      # Omitting the absolute path in this regex (.*)
+      And the output should match:
+      """
+      Util\.system! scp -C user@www\.example\.com:~\/dumps\/dump_for_download.dump .*?\/tmp\/aruba\/tmp\/staging.dump
+      """
+      And the output should contain "> Dumped the staging database to tmp/staging.dump"
```
Comment (codener) [inline features/dump.feature:50]: Please have this as a separate scenario. For one, it is a separate usage scenario :wink: and secondly, you accidentally got a wrong test green this way: L43 must contain "dumple staging primary" for the database selection, but the test is green because it sees the previous output of the first run.

Please also adapt the command output to reflect the different mode (include "primary"). You may omit steps that were already tested elsewhere and have no relevance for this very feature. Please also test the output when loading that dump (with `--load`). Suggestion:

```suggestion
Scenario: Creating a dump of one of multiple remote databases and loading it locally
  When I run `geordi dump staging --database primary --load`
    Then the output should contain "# Dumping the primary database of staging"
      And the output should contain "Util.system! ssh, user@www.example.com, -t, cd /var/www/example.com/current && bash --login -c 'dumple staging primary --for_download'"
      And the output should contain "> Dumped the primary staging database to tmp/staging.dump"
      
      # Loading the dump
      And the output should contain "Sourcing dump into the development db"
      And the output should contain "Your development database has now the data of staging (primary database)."
```

:information_source: you may tag scenarios with `@announce` (place that tag above `Scenario`) to see the output of the test run.

File: features/dump.feature [hunk 2]
```diff
@@ -6,6 +6,10 @@ Feature: The dump command
     Then the output should contain "Util.system! dumple development"
       And the output should contain "Successfully dumped the development database"
 
+    When I run `geordi dump -d primary`
+    Then the output should contain "Util.system! dumple development"
+      And the output should contain "Successfully dumped the development database"
+
```
Comment (codener) [inline features/dump.feature:12]: Please also have this as a separate scenario.
- check for "Dumping the primary development database"
- check for `Util.system! dumple development primary`
- check for "Successfully dumped the primary development database"

Comment (codener) [reply features/dump.feature:12]: You actually did not do any changes here yet. Please either add support for dumping multiple local databases, or remove that test.

File: lib/geordi/commands/dump.rb [hunk 3]
```diff
@@ -36,10 +42,12 @@ def dump(target = nil, *_args)
...
       loader = DumpLoader.new(options.load)
 
-      Interaction.announce "Sourcing dump into the #{loader.config['database']} db"
+      database = options[:database] ? " #{options[:database]}" : ''
+
+      Interaction.announce "Sourcing dump into the #{loader.config['database']}#{database} db"
       loader.load
 
-      Interaction.success "Your #{loader.config['database']} database has now the data of #{options.load}."
+      Interaction.success "Your #{loader.config['database']}#{database} database has now the data of #{options.load}."
```
Comment (codener) [inline lib/geordi/commands/dump.rb:50]: These changes do not seem correct to me. They claim a dump was loaded to eg. "development primary", but that is not true. Further, we're in the "no target" (= development) logic branch, which you did not even change.

You should instead change L59, remote.rb:50 and L68 as required by the updated tests.

File: features/dump.feature (hunk 1 above)
Comment (kajatiger) [reply features/dump.feature:50]: Ah okay thanks! I am not very familiar with cucumber steps 🙂

File: lib/geordi/commands/dump.rb (hunk 3 above)
Comment (kajatiger) [reply lib/geordi/commands/dump.rb:50]: I have to admit I don't fully understand what you meant in your first review with 
> Please add support for making a dump with multiple local databases by enhancing dump.rb:46
This was my confused interpretation of it.

Comment (codener) [reply lib/geordi/commands/dump.rb:50]: The dump command boils down to:
```
if target.nil? # No Capistrano stage given, so this run is about development
  if options.load # *Load* dump mode
     # load a dump file into the development database (1)
  else # Dump mode
    # make development dump (2)
else # Capistrano stage given (usually staging or production)
  # dump the given stage and download the dump (3)
  if options.load # loading the dump required
    # load the downloaded dump into the local database (3b)
  end
end
```

I was talking about the "make development dump" (2) part. When supporting to dump one of multiple remote databases (3), Geordi should equally be able to dump one of multiple local databases.

Comment (codener) [reply lib/geordi/commands/dump.rb:50]: > You should instead change L59, remote.rb:50 and L68 as required by the updated tests.

By this I meant that the output of step (3) and (3b) should be updated instead.

**Comment by codener** [comment]
Dear @kajatiger will you further elaborate on this PR, or should it be closed?

**Comment by kajatiger** [comment]
> Dear @kajatiger will you further elaborate on this PR, or should it be closed?

I have scheduled further work on this PR in my so called off-season at work (in 2 weeks) if that's okay 🙂

File: README.md [hunk 4]
```diff
@@ -169,6 +169,12 @@ When called without arguments, dumps the development database with `dumple`.
 
     geordi dump
 
+If you are using multiple databases per environment, pass the database name like this:
+
+    geordi dump -d primary
```
Comment (kajatiger) [inline README.md:174]: Add placeholder for primary or other database.

**Comment by codener** [comment]
@kajatiger Please let us know when you're done; best by re-requesting review.

File: features/dump.feature [hunk 5]
```diff
@@ -32,3 +37,22 @@ Feature: The dump command
       Util\.system! scp -C user@www\.example\.com:~\/dumps\/dump_for_download.dump .*?\/tmp\/aruba\/tmp\/staging.dump
       """
       And the output should contain "> Dumped the staging database to tmp/staging.dump"
+
+  Scenario: Creating a dump of one of multiple remote databases and loading it locally
+    Given a file named "Capfile" with "Capfile exists"
+    And a file named "config/deploy.rb" with:
+    """
+    """
```
Comment (codener) [inline features/dump.feature:45]: Please also have this as `a file named "config/deploy.rb" with "deploy.rb exists"`.

File: features/dump.feature [hunk 6]
```diff
@@ -32,3 +37,22 @@ Feature: The dump command
...
+    set :deploy_to, '/var/www/example.com'
+    set :user, 'user'
+
+    server 'www.example.com'
+    """
+
+    When I run `geordi dump staging --database primary`
+    Then the output should contain "# Dumping the primary database of staging"
+      And the output should contain "Util.system! ssh, user@www.example.com, -t, cd /var/www/example.com/current && bash --login -c 'dumple staging primary --for_download'"
+      And the output should contain "> Dumped the primary staging database to tmp/staging.dump"
```
Comment (codener) [inline features/dump.feature:58]: I am missing the `--load` option in L55 and the tests for the expected loading behavior (as discussed):

```
      # Loading the dump
      And the output should contain "Sourcing dump into the development db"
      And the output should contain "Your development database has now the data of staging (primary database)."
```

File: lib/geordi/commands/dump.rb [hunk 7]
```diff
@@ -5,18 +5,35 @@
...
 
+
 `geordi dump staging -l` (with a Capistrano deploy target and the `--load`
 option) sources the dump into the development database after downloading it.
 DESC
 
+
+option :load, aliases: ['-l'], type: :string, desc: 'Load a dump'
+option :select_server, default: false, type: :boolean, aliases: '-s'
+option :database, aliases: '-d', type: :string, desc: 'Specify key, under which database is nested in case of multiple', banner: 'DATABASE'
```
Comment (codener) [inline lib/geordi/commands/dump.rb:29]: Please remove the duplicate load option and the select_server option. I assume they were created during a rebase.

File: lib/geordi/commands/dump.rb [hunk 8]
```diff
@@ -33,21 +50,21 @@ def dump(target = nil, *_args)
...
   else
-    Interaction.announce 'Dumping the database of ' + target
+    Interaction.announce "Dumping the #{database}database of #{target}"
     dump_path = Geordi::Remote.new(target).dump(options)
 
     if options.load
       loader = DumpLoader.new(dump_path)
 
-      Interaction.announce "Sourcing dump into the #{loader.config['database']} db"
+      Interaction.announce "Sourcing dump into the #{database}#{loader.config['database']} db"
```
Comment (codener) [inline lib/geordi/commands/dump.rb:64]: This line should not contain the `#{database}` fragment, as you did not implement sourcing a dump into multiple databases. It is always the single `development` database.

File: lib/geordi/commands/dump.rb [hunk 9]
```diff
@@ -33,21 +50,21 @@ def dump(target = nil, *_args)
...
       loader.load
 
-      Interaction.success "Your #{loader.config['database']} database has now the data of #{target}."
+      Interaction.success "Your #{loader.config['database']} database has now the data of #{target}#{database}."
```
Comment (codener) [inline lib/geordi/commands/dump.rb:67]: Your change prints "... the data of stagingprimary." Please print "... the data of staging (primary database)." instead, as discussed.

File: lib/geordi/commands/dump.rb (hunk 7 above)
Comment (codener) [inline lib/geordi/commands/dump.rb:29]: Please change the description:

```suggestion
option :database, aliases: '-d', type: :string, desc: 'Database name, if there are multiple databases', banner: 'NAME'
```

File: lib/geordi/commands/dump.rb [hunk 10]
```diff
@@ -5,18 +5,35 @@
...
+
+If you are using multiple databases per environment, pass the database name like this:
+
+    geordi dump -d primary
+
+Loading a dump into one of multiple local databases is not supported yet.
+
+When called with the `--load` option, sources the specified dump into the
+development database.
+
```
Comment (codener) [inline lib/geordi/commands/dump.rb:17]: I assume these lines got mislocated during a rebase. Please move them to the very end of the description (right before L24) and drop L15+16, which seems like a rebase relict as well.

File: README.md [hunk 11]
```diff
@@ -403,6 +417,13 @@ Copy `lib/geordi/COMMAND_TEMPLATE` to `lib/geordi/commands/your_command` and
 edit it to do what you need it to do. Please add a feature test for the new
 command; see features/ for inspiration.
 
+
+    # @option -I: add directory to load path
+    ruby -I ../geordi/lib ../geordi/bin/geordi <command>
+
+    # with debugger
+    ruby -r byebug -I ../geordi/lib ../geordi/bin/geordi <command>
+
```
Comment (codener) [inline README.md:426]: Maybe this also was added during a rebase. Please remove or discuss :wink:

File: README.md [hunk 12]
```diff
@@ -416,3 +437,4 @@ To try Geordi locally, call it like this:
 
 You can also *install* Geordi locally from its project directory with
 `rake install`. Make sure to switch to the expected Ruby version before.
+
```
Comment (codener) [inline README.md:440]: Please remove this trailing newline.

File: README.md [hunk 13]
```diff
@@ -170,9 +170,23 @@ offer to edit the whitelist instead.
...
+    geordi dump -d <name of key where db is nested>
+
+Ususally in Rails the first database is nested under `primary`.
+
+Loading a dump into one of multiple local databases is not supported yet.
+
+When called with the `--load` option, sources the specified dump into the
+development database.
+
+
```
Comment (codener) [inline README.md:185]: This looks mislocated as well. Please run `rake readme` and check if it fixed this.

**Review by codener — CHANGES_REQUESTED** [review]
We're approaching the finish line. Please make sure to install your version of Geordi locally (`rake install`) and try it. When you've updated the PR, I will install and try it as well. Once it is working fine, I'll release.

Best regards

**Comment by codener** [comment]
Dear @kajatiger do you still intend to finish this? Else I'll close the PR.

**Comment by kajatiger** [comment]
> Dear @kajatiger do you still intend to finish this? Else I'll close the PR.

Yes I do want to finish this. As I said I can't always work on it cause of work priorities, but will hopefully finalize today.

File: features/dump.feature (hunk 6 above)
Comment (kajatiger) [reply features/dump.feature:82]: We realized that there was no feature test for the load option in general, so we implemented that as well.

File: lib/geordi/commands/dump.rb [hunk 14]
```diff
@@ -8,15 +8,26 @@
...
 `geordi dump staging -l` (with a Capistrano deploy target and the `--load`
 option) sources the dump into the development database after downloading it.
```
Comment (codener) [inline lib/geordi/commands/dump.rb:21]: Please remove the "When called with..." sentence as discussed, and please move L20+21 ("geordi dump staging -l (with a Capistrano deploy target ...") below "geordi dump staging", atop your insertion. I'd like your addition at the bottom of `DESC`.

**Review by codener — CHANGES_REQUESTED** [review]
Thanks for the update. Please have a look at the command description. When you're done, please run `rake readme` a last time. It will update the README from all command descriptions and e.g. add the `database` option to the `dump` section in the README.

Meanwhile I'll try it locally.

**Comment by codener** [comment]
So I tried Geordi with your addition. It works well :+1: 
I have two more wishes:

* Please include the database name in the dumple output (if present) https://github.com/makandra/geordi/blob/0cea55258e7c34e895d3eb419276a7c4f7fd5341/exe/dumple#L65
* Please provide some simple error handling if someone requests a nested database, but misspells the name, or multiple databases do not even exist. E.g. by changing this https://github.com/makandra/geordi/blob/0cea55258e7c34e895d3eb419276a7c4f7fd5341/exe/dumple#L51 to
  ```
  if database
    config = config[database] or raise %(Unknown #{environment} database "#{database}")
  end
  ```

**Comment by kajatiger** [comment]
Okay, I hope we are all good to go 🙂

File: exe/dumple [hunk 15]
```diff
@@ -62,7 +65,7 @@ begin
     dump_path = "#{dump_dir}/#{config['database']}_#{Time.now.strftime("%Y%m%d_%H%M%S")}.dump"
   end
 
-  puts "> Dumping database for \"#{environment}\" environment ..."
+  puts "> Dumping \"#{config['database'] + ' '}\"database for \"#{environment}\" environment ..."
```
Comment (codener) [inline exe/dumple:68]: Ok, this was ambiguous, but for consistency with the rest of Geordi the value of the `database` variable should be used:

```suggestion
  given_database = database ? %(#{database} ) : ""
  puts "> Dumping #{given_database}database for \"#{environment}\" environment ..."
```

**Review by codener — APPROVED** [review]
I added one update, when it's applied I'll approve.

As a side note: Please do not squash commits during repeated code reviews. It makes it really hard to follow changes as the reviewer is presented the whole changeset each time. The right time for squashing is _after_ approval. Thank you.

**Comment by codener** [comment]
FYI we'll have a release on Friday. When this PR is ready, it will be included.

**Review by codener — APPROVED** [review]

**Comment by brunosedler** [comment]
Released in 4.2.0
